src/de.rs

```diff
@@ -44,6 +44,52 @@ where
     Ok(t)
 }
 
+/// Deserializes a `&[u8]` into a type.
+///
+/// This function will attempt to interpret `bytes` as the BCS serialized form of `T` and
+/// deserialize `T` from `bytes`.
+///
+/// # Safety
+///
+/// This is unsafe. Developers must ensure that their `Deserialize`
+/// structs correctly reflect the data being deserialized.
+/// If the deserialization logic of one type overlaps with the next field,
+/// the overall deserialization might succeed with undefined behavior.
+///
+/// If you're not sure what you're doing, consider using `from_bytes` instead.
+///
+/// # Examples
+///
+/// ```
+/// use bcs::from_bytes_discarding_remaining_input;
+/// use serde::Deserialize;
+///
+/// #[derive(Deserialize)]
+/// struct Ip([u8; 4]);
+///
+/// #[derive(Deserialize)]
+/// struct SocketAddr {
+///     ip: Ip,
+/// }
+///
+/// let bytes = vec![0x7f, 0x00, 0x00, 0x01, 0x41, 0x1f];
+/// let socket_addr: SocketAddr = unsafe {
+///     from_bytes_discarding_remaining_input(&bytes).unwrap()
+/// };
+///
+/// assert_eq!(socket_addr.ip.0, [127, 0, 0, 1]);
+/// ```
+#[allow(unsafe_code)]
+pub unsafe fn from_bytes_discarding_remaining_input<'a, T>(bytes: &'a [u8]) -> Result<T>
```

Comment:
I wouldn't use `unsafe` for that. I believe what you're asking is a variant where we don't call `end()`. Or equivalently exposing the `Deserializer` struct. Historically, @bmwill and I have pushed back on this idea. Not sure if there is a stronger argument now.